File: python/learn_unlearn.py
from config import *
import numpy
from numpy import *
from tools import *

def learn_unlearn(n_map, n_iter, center, patch, in_param, common):
    '''
    Timothee Masquelier timothee.masquelier@alum.mit.edu Nov 2006
    n_map: number of input maps
    center: eventually previous centers (to go on a computation), otherwise
    leave empty: []
    patch: eventually a patch to begin with (instead of random weights),
           otherwise leave empty: []
    rec_filter: filter to use for reconstructions
    PARAM structure:
    PARAM.n: number of cells
    PARAM.RFSize: RF size of cells
    PARAM.kWTA: limit the number of cells that can learn with a given image
    '''
    
    #Set random seed for test mode
    if params['test_mode']:
        random.mtrand.seed(5489)

    params['s']['normalize'] = False
    params['s']['learning_period'] = 200.0
    tmp = get_onoff_map(0, common)

    n_filter = shape(tmp)[2]

    if center == []:
        has_initial_path = (patch != [])
        #instantiate main structure
        for cent_ind in range(in_param['n']):
            if not has_initial_path:
                patch.append(get_random_centers(1, in_param['rf_size'],
                             n_filter))
                center.append({})
                center[cent_ind]['original_patch'] = patch[cent_ind]
                patch[cent_ind] = patch[cent_ind].flatten()
            else:
                raise NameError('Error')


            center[cent_ind]['activity'] = numpy.float64(-1.0)
            center[cent_ind]['n_firing'] = 0.0
            center[cent_ind]['evol_theta_original'] = zeros(n_iter,
                                                            numpy.float64)
            center[cent_ind]['evol_threshold'] = zeros(n_iter, numpy.float64)
            center[cent_ind]['activity_threshold'] = numpy.float64(0.0)
            center[cent_ind]['learning_rate'] = in_param['learning_rate_max']


    trace = None
    if in_param['nu'] < inf:
        trace = 1.0 / in_param['nu'] * ones(in_param['n'], numpy.float64)
    else: #mode with no trace
        trace = ones(in_param['n'], numpy.float64)


    neutral = ones((in_param['rf_size'][0], in_param['rf_size'][1], n_filter),
                   numpy.float64)
    neutral = neutral.flatten(1) / linalg.norm(neutral.flatten(1))

    im_order = permute(n_map) 
    current_im_idx = 0

    print 'lenarnUnlearn: size ', str(in_param['rf_size']), '...'

    
    #Iteration-------------------------------------------------
    for iteration in range(n_iter):
        #if iteration == 38115:
        #    pdb.set_trace()

        onoff_map = get_onoff_map(im_order[current_im_idx], common)
        for cent_ind in range(in_param['n']):
            img = onoff_map.flatten(1)
            #this normalization is very useful (Tim 04/2007)
            img = img / linalg.norm(img)

            center[cent_ind]['raw_activity'] = dot(img, patch[cent_ind])
            center[cent_ind]['activity'] = center[cent_ind]['raw_activity'] / \
                trace[cent_ind]
            
        #inhibition
        #apply inhibition to raw inputs
        for cent_ind in range(in_param['n']): #affect inhibited activites
            center[cent_ind]['activity'] = \
            solve_nonselect_inhib_circuit(center[cent_ind]['activity'],
                                              in_param['inhib'])
            
         
        max_activity = float64(0.0)
        most_active = 0
        for cent_ind in range(in_param['n']):
            if center[cent_ind]['activity'] > max_activity:
                max_activity = center[cent_ind]['activity']
                most_active = cent_ind

       
        #Check for fire---       
        if max_activity >= center[most_active]['activity_threshold']:
            
            center[most_active], patch[most_active] = fire(center[most_active],
                                                           patch[most_active],
                                                           img, iteration,
                                                           in_param)
            
            if in_param['push_away']:
                for cent_ind in range(in_param['n']):
                    if cent_ind == most_active:
                        continue
                    if center[cent_ind]['activity'] >= \
                        center[cent_ind]['activity_threshold']:
                        patch[cent_ind] = patch[cent_ind] + 1.0 * \
                        center[cent_ind]['learnng_rate'] * 1.0 * \
                        (neutral - patch[cent_ind])
                        if in_param['normalize']:
                            patch[cent_ind] = patch[cent_ind] / \
                            linalg.norm(patch[cent_ind])


        #Update traces
        for cent_ind in range(in_param['n']):
            trace[cent_ind] = center[cent_ind]['raw_activity'] / in_param['nu']\
            + float64(1.0 - 1.0 / in_param['nu']) * trace[cent_ind]

        #Update threshold
        for cent_ind in range(in_param['n']):
            center[cent_ind]['activity_threshold'] = \
            center[cent_ind]['activity_threshold'] * \
            float64(1.0 - in_param['thr_decay'])

        if current_im_idx == n_map -1: #last image
            current_im_idx = 0
            im_order = permute(n_map) #repermute
        else:
            current_im_idx += 1 #move on to next image

        if mod(iteration, round(n_iter / 100.0)) == 0:
            print '.',

        if iteration == 599:
            tsum = 0            
            for cent_ind in range(in_param['n']):
                tsum += center[cent_ind]['n_firing']
            if tsum == 0:
                print 'Neurons do not fire'
                break

        

    #-End of oiteration ---------------------------


    for cent_ind in range(in_param['n']):
        center[cent_ind]['patch'] = reshape(patch[cent_ind],
                                            (in_param['rf_size'][0],
                                            in_param['rf_size'][1], n_filter),
                                            order='F')

    return center



#---Fire Function--------------------------------------------------
def fire(center, patch, best_patch, iteration, in_param):
    '''Fire the neuron'''

    #inc firing_rate
    center['n_firing'] += 1

    #Hebbian modification: dW = a.y.(X-W)
    y_var = float64(1.0)
    
    new_center_patch = patch + (center['learning_rate'] * y_var * \
                                (best_patch - patch)).astype(numpy.float64)

    #Normalize
    if in_param['normalize']:
        new_center_patch = new_center_patch / linalg.norm(new_center_patch)
        center['evol_theta_original'][iteration] = \
        arccos(minimum(1, dot(new_center_patch, center['original_patch'])))
    else:
        center['evol_theta_original'][iteration] = \
        arccos(minimum(1, dot(new_center_patch / linalg.norm(new_center_patch),
               center['original_patch'].flatten())))

    if (center['n_firing'] <= in_param['learning_period']) and \
        mod(center['n_firing'], 10) == 0:
        center['learning_rate'] = float64(center['learning_rate'] * \
                                          (in_param['learning_rate_min'] / \
                                          in_param['learning_rate_max']) ** \
                                          (1.0 / (in_param['learning_period'] /\
                                          10.0)))

    center['activity_threshold']  = in_param['thr_coef'] * center['activity']
    center['evol_threshold'][iteration] = center['activity_threshold']

    return center, new_center_patch

#-Get Random Centers Function ----------------------------------------
def get_random_centers(n_size, rf_size, n_filter):
    '''return a random center matrix'''

    centers = zeros((rf_size[0], rf_size[1], n_filter, n_size), dtype=float64)

    for cent_ind in range(n_size):
        center = float64(random.random((rf_size[0], rf_size[1], n_filter)))
       
        center = maximum(0, center) 
        centers[:, :, :, cent_ind] = center / linalg.norm(center.flatten(1))

    return centers


#- Solve non selective inhib circuit function ----------
def solve_nonselect_inhib_circuit(input_data, alpha):
    '''Solve non selective inhibit circuits'''

    if alpha == 0:
        return input_data
    i_var = alpha * mean(maximum(0, input_data-0)) - 0
    y_var = maximum(0, input_data - i_var)

    return y_var



'''if __name__ == "__main__":
    common = {}
    common['on_off_map'] = load(paths['map'] + 'onoff.' + str(0) + '.' \
                                + str(0) + '.npy')
    center = learn_unlearn(len(common['on_off_map']), params['s']['n_iter'], [],
                           [], params['rec_filter'], params['s'], common)'''

File: matlab/learnUnlearn.m
function [center] = learnUnlearn(nMap,nIter,getMap,center,patch,recFilter,PARAM)
% Timothee Masquelier timothee.masquelier@alum.mit.edu Nov 2006
% nMap: number of input maps
% getMap: handle to a function i -> map
% center: eventually previous centers (to go on a computation), otherwise
% leave empty: []
% patch: eventually a patch to begin with (instead of random weights), otherwise leave empty: []
% recFilter: filter to use for reconstructions
% PARAM structure:
%   PARAM.n: number of cells
%   PARAM.RFSize: RF size of cells
%   PARAM.kWTA: limit the number of cells that can learn with a given image

rand('twister',5489);
%inhib
% PARAM.inhibWeight = 0.5;% horizontal inhibition
% PARAM.alpha = .01;
% PARAM.beta = .001;
% PARAM.gamma = .01;
% PARAM.lambda = 10;% sigmoid parameter
% PARAM.p = 1/PARAM.n;
PARAM.learningPeriod = 200;
PARAM.normalize = false; % not necessary to normalize weight (Tim 04/2007)
% PARAM.nu = 100;

tmp = feval(getMap,1);
nFilter = size(tmp{1},3);


if isempty(center)
    hasInitialPatch = ~isempty(patch);
    %instantiate main structure
    for iCenter=1:PARAM.n
        if ~hasInitialPatch
            patch{iCenter} = getRandomCenters(1,PARAM.RFSize,nFilter);
            center(iCenter).originalPatch = patch{iCenter};
            patch{iCenter} = patch{iCenter}(:); % for faster computations            
        else
            error('TO DO')
            center(iCenter).patch = reshape(patch(:,iCenter),[PARAM.RFSize nFilter]);
            center(iCenter).patch = center(iCenter).patch ./ norm(center(iCenter).patch(:));
        end
%         center(iCenter).minDistance = Inf;
        center(iCenter).activity = -1;
        %         center(iCenter).minBand = -1;
        %         center(iCenter).minI = -1;
        %         center(iCenter).minJ = -1;
        %         center(iCenter).distanceMap = {};
        center(iCenter).nFiring = 0;
        %         center(iCenter).evolDist = zeros(1,nIter);
        center(iCenter).evolThetaOriginal = zeros(1,nIter);
        center(iCenter).evolThreshold = zeros(1,nIter);
        %         center(iCenter).evolDeltaTheta = zeros(1,nIter);
        %         center(iCenter).evolTheta = zeros(1,nIter);
        %         center(iCenter).evolBand = zeros(1,nIter);
                center(iCenter).activityThreshold = 0;
        center(iCenter).learningRate = PARAM.learningRateMax;
    end
end

if PARAM.nu < Inf
    trace = 1/PARAM.nu*ones(1,PARAM.n);
else % mode with no trace
    trace = ones(1,PARAM.n);
end    

% for now
% threshold = zeros(1,PARAM.n);

        neutral = ones([PARAM.RFSize nFilter]);
        neutral = neutral(:)/ norm(neutral(:));


imOrder = randperm(nMap);
currentImIdx = 1;

idx = 1:PARAM.n;

disp(['learnUnlearn: size ' int2str(PARAM.RFSize) ' ...'])

tic

for iter=1:nIter
    
%     if mod(iter-1,100000)==0
%         %         fprintf(1,'.',iter);
%         for iCenter=1:PARAM.n
%             center(iCenter).patch = reshape(patch{iCenter}, [PARAM.RFSize nFilter]);
%         end
%         rec(center,recFilter,['iter ' int2str(iter-1)])
%     end

    map = getMap(imOrder(currentImIdx));
    %     map = feval(getMap,imOrder(currentImIdx));

    for iCenter=1:PARAM.n
        im = map{1};
        im = im(:);
        im = im / norm(im); % this normalization is very useful (Tim 04/2007)
        
        center(iCenter).rawActivity = dotProduct(im,patch{iCenter});
        center(iCenter).activity = center(iCenter).rawActivity/trace(iCenter);
%         if ~PARAM.normalize % note: if PARAM.normalize then this operation is unuseful, otherwise it helps balacing the competition between neurons
%                 center(iCenter).activity = center(iCenter).activity/norm(patch{iCenter});
%         end
    end
    
    % inhibition
%     apply inhibition to raw inputs
%     y = solveNonSelectiveMaxInhibCircuit([center.activity],PARAM.inhib);
    y = solveNonSelectiveInhibCircuit([center.activity],PARAM.inhib);
    for iCenter=1:PARAM.n % affect inhibited activites
        center(iCenter).activity = y(iCenter);
    end
    

    [maxActivity mostActive] = max([center.activity]);
    
    maxRawActivity = center(mostActive).rawActivity;

    if maxActivity >= center(mostActive).activityThreshold
%     if maxRawActivity >= center(mostActive).activityThreshold
        [center(mostActive) patch{mostActive}] = fire(center(mostActive), patch{mostActive}, im,iter,PARAM);
        
%         % reconstruction
%         center(mostActive).patch = reshape(patch{iCenter}, [PARAM.RFSize nFilter]);
%         rec(center(mostActive),recFilter,['iter ' int2str(iter-1)])
%     
%         if center.nFiring==50
%             error('stop')
%         end
        
        if PARAM.pushAway
            % push away other neurons
            for iCenter=1:PARAM.n
                if iCenter==mostActive
                    continue
                end
                if center(iCenter).activity >= center(iCenter).activityThreshold % would have fired later in the abscence of 1WTA
                    %                 newPatch = patch{iCenter} - center(iCenter).learningRate * 1 * ( im/norm(im(:)) - patch{iCenter} );
                    patch{iCenter} = patch{iCenter} + 1 * center(iCenter).learningRate * 1 * ( neutral - patch{iCenter} );
                    if PARAM.normalize
                        patch{iCenter} = patch{iCenter} / norm(patch{iCenter});
                    end
                end
            end
        end
    end
    
    % update traces
%     trace = [center.activity] / PARAM.nu + (1-1/PARAM.nu) * trace;
    trace = [center.rawActivity] / PARAM.nu + (1-1/PARAM.nu) * trace; % seems to be a better idea (Tim 04/2007)
    
%     for iCenter=1:PARAM.n % update thr for neuron that could have fired
%         if center(iCenter).activity>center(iCenter).activityThreshold
% %         if center(iCenter).rawActivity>center(iCenter).activityThreshold
%             center(iCenter).activityThreshold = PARAM.thrCoef*center(iCenter).activity;
% %             center(iCenter).activityThreshold = PARAM.thrCoef*center(iCenter).rawActivity;
%         end
%     end
        


    % update thr
        for iCenter=1:PARAM.n
            center(iCenter).activityThreshold = center(iCenter).activityThreshold * (1-PARAM.thrDecay);
        end


    %     fprintf(1,'.');
    if currentImIdx == nMap % last image
        currentImIdx = 1;
        imOrder = randperm(nMap); % reshuffle
    else
        currentImIdx = currentImIdx+1; % move on to next image
    end
    %             fprintf(1,'\n');



        if mod(iter,round(nIter/100))==0
            fprintf(1,'.');
        end
    if iter==600 && sum([center.nFiring])==0
        disp('Neurons do not fire.')
        break;
    end

end % iteration

fprintf(1,'\n');
toc

for iCenter=1:PARAM.n
    center(iCenter).patch = reshape(patch{iCenter}, [PARAM.RFSize nFilter]);
end

function [center newCenterPatch] = fire(center,patch,bestPatch,iter,PARAM)
% RFSize = size(bestPatch,1);
% alpha = (PARAM.size/4)^2; % see Mutch CVPR 2006
alpha = 1;

% inc firing counts
center.nFiring = center.nFiring+1;

% Hebbian modification: dW = a.y.(X-W)
y = 1;
% y = (1-center.minDistance/PARAM.firingThreshold/PARAM.size)^1;
% y = exp( -center.minDistance / 2 / center.sigma^2 / alpha );
% y = max(0,1/center.d2Threshold*(center.d2Threshold-center.minDistance));
% y = y^2 * 3;
% newCenterPatch = patch + PARAM.learningRate * y * bestPatch;
newCenterPatch = patch + center.learningRate * y * ( bestPatch - patch );
% newCenterPatch = patch + center.learningRate * y * ( bestPatch - patch );

% normalize
if PARAM.normalize
    newCenterPatch = newCenterPatch / norm(newCenterPatch);
end

% Reporting
% % compute delta theta
% center.evolDeltaTheta(iter) = acos(min(1,dotProduct(patch(:),newCenterPatch(:))));
% % compute Theta
% center.evolTheta(iter) = acos(min(1,dotProduct(patch(:),normedBestPatch(:))));
% % dist
% center.evolDist(iter) = center.maxActivity;
% % band
% center.evolBand(iter) = center.minBand;
% % dist
if PARAM.normalize
    center.evolThetaOriginal(iter) = acos(min(1,dotProduct(newCenterPatch,center.originalPatch)));
else
    center.evolThetaOriginal(iter) = acos(min(1,dotProduct(newCenterPatch/norm(newCenterPatch),center.originalPatch)));
end

% % store new center
% patch = newCenterPatch;
% 
% compute new threshold if necessary
if center.nFiring <= PARAM.learningPeriod && mod(center.nFiring,10)==0
    % geometric
    center.learningRate = center.learningRate * (PARAM.learningRateMin/PARAM.learningRateMax)^(1/(PARAM.learningPeriod/10));
    %     center.d2Threshold = center.d2Threshold * (PARAM.d2ThresholdMin/PARAM.d2ThresholdMax)^(1/(PARAM.learningPeriod/10));
    % arithmetic
    %     center.learningRate = center.learningRate - (PARAM.learningRateMax-PARAM.learningRateMin)/(PARAM.learningPeriod/10);
    %     center.d2Threshold = center.d2Threshold - (PARAM.d2ThresholdMax-PARAM.d2ThresholdMin)/(PARAM.learningPeriod/10);

    %     center.activityThreshold = exp(-center.d2Threshold / (2*PARAM.sigma^2*alpha));
end
% center.activityThreshold = center.maxActivity * center.activityThreshold ;

center.activityThreshold = PARAM.thrCoef*center.activity;
% center.activityThreshold = PARAM.thrCoef*center.rawActivity;

% center.activityThreshold = center.activity;
center.evolThreshold(iter) = center.activityThreshold;

% function center = computeMaxActivity(center,sigma,alpha)
%
% center.minDistance = Inf;
% for iBand=1:length(center.distanceMap)
%     [distance idx] = min(center.distanceMap{iBand}(:));
%     if distance < center.minDistance % found a better match
%         center.minDistance = max(0,distance);% just to avoid tiny negative values
%         center.maxActivity = exp( -distance / 2 / (sigma^2) / alpha );
%         center.minBand = iBand;
%         center.minJ = floor((idx-1)/size(center.distanceMap{iBand},1))+1;
%         center.minI = idx - (center.minJ-1) * size(center.distanceMap{iBand},1);
%     end
% end

function centers = getRandomCenters(n,RFSize,nFilter)
MEAN = 20;
STD = 1;

centers = [];

for iCenter=1:n
    %     center = MEAN + STD * randn([PARAM.size PARAM.size nFilter]);
    %         center = rand([PARAM.size PARAM.size nFilter]);
    %     center = ones([PARAM.size PARAM.size nFilter]);

    center = 0*1+1*rand([RFSize nFilter]);

    %     sigma = RFSize(1)/4;
    %     x = abs( [1:RFSize(1)] - (RFSize(1)+1)/2 )' * ones(1,RFSize(2));
    %     y = abs( [1:RFSize(2)] - (RFSize(2)+1)/2 )' * ones(1,RFSize(1));
    %     y = y';
    %     g = exp(-( x.^2 + y.^2 )/2/(sigma^2));
    %
    %     for n=1:nFilter
    %         center(:,:,n) = g;
    %     end
    %
    center = max(0,center); % rectify negative values
    centers(:,:,:,iCenter) = center ./ norm(center(:));
end

